### ISSUE TITLE
WCS `pix2world` does not change sign of HPLN-TAN

### ISSUE THREAD

**Issue by tiagopereira** [astropy/astropy#7716]
When using WCS with Helioprojective coordinates (HPLN-TAN and HPLT-TAN), the pixel to world coordinates is unable to change the sign of the longitude angle. It returns the conjugate angle instead, using the sign defined in CRVAL.

An example:

```
test = {'CTYPE1': 'HPLN-TAN', 'CUNIT1': 'deg', 'CDELT1': 1.,
        'CRPIX1': 5, 'CRVAL1': 1., 'NAXIS1': 10,
        'CTYPE2': 'HPLT-TAN', 'CUNIT2': 'deg', 'CDELT2': 1.,
        'CRPIX2': 5, 'CRVAL2': 1., 'NAXIS2': 10}
w = wcs.WCS(test)
w.all_pix2world(np.array([[0, 0]]), 1)
```

This returns:

    array([[356.0194272 ,  -3.97235806]])

But I would have expected `[-4, -4]`. It properly converted the angle for the latitude, but for longitude it gives 360 + angle.

If I create the same coordinate system but using a different CRVAL, you can see it works properly for the first pixel:

```
test = {'CTYPE1': 'HPLN-TAN', 'CUNIT1': 'deg', 'CDELT1': 1.,
        'CRPIX1': 1, 'CRVAL1': -3., 'NAXIS1': 10,
        'CTYPE2': 'HPLT-TAN', 'CUNIT2': 'deg', 'CDELT2': 1.,
        'CRPIX2': 5, 'CRVAL2': 1., 'NAXIS2': 10}
w = wcs.WCS(test)
w.all_pix2world(np.array([[0, 0]]), 1)
# array([[-3.99853007, -3.98676173]])
```
So it gives the same sign for HPLN-TAN as defined in `CRVAL1`. If I try to get a coordinate on the positive side, it will give again the (negative) conjugate angle:

```
w.all_pix2world(np.array([[8, 0]]), 1)
# array([[-356.04396038,   -3.95810938]])
```

This should not be happening. The coordinates should be negative or positive, not conjugate angles to keep the same sign as CRVAL.

**Comment by nden** [comment]
I am not sure I understand what problem you are pointing out. Longitude values are in the range 0 - 360 deg. I've gone through the wcslib code and everything looks consistent and the results you report are expected. Specifying a valid WCS is left to the user. if you choose to specify negative CRVAL for the longitude then the output is aligned with that choice. So unless there's something specific to HP-frames, which I am not familiar with, the results look correct to me.

**Comment by tiagopereira** [comment]
HP-frames, widely used in solar physics, and known as solar (x, y) coordinates (x for longitude, y for latitude), have the centre of the projected solar disk at (0, 0). Therefore, any longitude that is on the eastern hemisphere is negative, and longitudes on the western hemisphere are positive. The same for latitude and northern/southern hemispheres. Because the whole Sun is a lot less wide than one degree, getting longitudes of hundreds of degrees sounds ridiculous. Typically in these coordinates pointings are always within [-1000", 1000"].

The error does not arise when supplying invalid WCS. In fact, many space missions today use these coordinates and I first saw it when using their own FITS files.

I also don't see why you single out longitude, when it seems that latitude works fine with both positive and negative.

**Comment by Cadair** [comment]
I believe this is just an issue that wcslib assumes all longitude values are in the range 0 - 360 and not -180 - 180. I was under the impression that the sign of CRVAL does not affect this assumption at all.

**Comment by Cadair** [comment]
> The error does not arise when supplying invalid WCS

I believe this assumption is introduced by there being a known projection in the CTYPE.

**Comment by tiagopereira** [comment]
I also noticed that WCSAxes doesn't seem to have a problem with these WCS data and plots the longitude correctly, allowing changes from negative to positive in the same image.

**Comment by Cadair** [comment]
Yes, that's because WCSAxes has it's own handling of longitude wrap angles, and I added the 180 deg wrap angle for the HPLN/T frames.

@nden may be able to correct me, but I was under the impression that making WCS/wcslib aware of wrap angle for longitude would not be trivial at all. There are a few places where WCS/wcslib really messes up solar frames, (like not knowing about observer location), we have generally just hacked around them in SunPy.

**Comment by nden** [comment]
> all longitude values are in the range 0 - 360 and not -180 - 180

Yes, wcslib treats all spherical systems the same way, i.e. the valid range for longitude is 0 to 360 and not -180 to 180.  The range for latitude is -90 to 90 , which is why I didn't mention it before, I thought we agree the values are correct. I believe negative longitude is allowed. Internally wcslib requires (or maybe assumes) the system sis right-handed and the onus is on the WCS designer to provide a correct WCS definition.
